### ISSUE TITLE
Improve Key Tap Accuracy Based on Probabilistic Models

### ISSUE THREAD

**Issue by MatthewScholefield** [futo-org/android-keyboard#433]
### Use Cases

I have a much higher typo rate when typing on FUTO keyboard than on GBoard. Upon closer analysis (and based on some video I watched that I can't find a link to), GBoard appears to scale the key area boundaries based on the probability of the letter appearing next in the current word / sentence. For example, I use the following two test sentences:

 - Sentence 1: `I'm going`
 - Sentence 2: `I will finis`

When I click within the word gap between "g" and "h" on GBoard, I consistently get the correct letter for both sentence 1 and sentence 2 indicating that it knows which letter is more probable and uses that to adjust the boundaries between letters (enabling "Show taps" in developer settings makes this even more obvious). When replicating the experiment on FUTO keyboard I consistently get a 50/50 split between the two letters indicating no such intelligence.

**TL;DR, has FUTO considered some sort of dynamic word boundary scaling to improve touch accuracy within word boundaries?**

### Proposal

Some sort of dynamic key boundary scaling based on likelihood of the next letter given the current word / sentence.

The model could just be a simple markov chain baked into the source code on a per-language basis. You'd want to scale the boundaries in a way that covers the gaps between words but still leaves the centers of words as their own letter so as to not cause frustration when typing specific characters.

### References

 -  futo-org/android-keyboard#682: If the areas between letters had a much higher probability of being the correct letter then perhaps it would be less important to configure deadzones between these letters.

**Comment by abb128** [comment]
Something like this was implemented in 0.1.24.1 by boosting letters that will lead to a valid dictionary word
